### PR TITLE
RavenDB-19165 Allow run RavenDB without admin privilage

### PR DESCRIPTION
**Pull request by haludi**
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-19165/Loading-the-certificate-requires-admin-permissions

### Additional description

In the past, we had an issue loading certificates when the application doesn't load the user's profile.
To solve that we changed the default key set (`UserKeySet` usually) to `MachineKeySet`.
MachineKeySet works well unless the process doesn't have sufficient privilege.
To deal with both cases we will load certificates with UserKeySet and failover to `MachineKeySet`.

_Please delete below the options that are not relevant_

### Type of change
- Bug fix

### How risky is the change?
- High

### Backward compatibility
- Non breaking change

### Is it platform specific issue?
- No

### Documentation update
- No documentation update is needed 

### Testing 
- It has been verified by manual testing

### Is there any existing behavior change of other features due to this change?
- No

### UI work
- No UI work is needed

### CODE REVIEW

File: src/Raven.Server/Utils/CertificateUtils.cs [hunk 1]
```diff
@@ -209,15 +244,15 @@ public static X509Certificate2 CreateSelfSignedClientCertificate(string commonNa
...
+            var cert = CreateCertificate(certBytes, flags: FlagsForPersist);
             return cert;
         }
 
         private static void ValidateNoPrivateKeyInServerCert(byte[] serverCertBytes)
         {
             var collection = new X509Certificate2Collection();
             // without the server private key here
-            collection.Import(serverCertBytes, (string)null, X509KeyStorageFlags.MachineKeySet);
+            collection.Import(serverCertBytes, (string)null, FlagsForExport);
```
Comment (ppekrol) [inline src/Raven.Server/Utils/CertificateUtils.cs]: shouldn't this have FlagsForOpen?

File: src/Raven.Client/Util/TcpUtils.cs [hunk 2]
```diff
@@ -204,8 +204,11 @@ Logger log
             if (info.Certificate == null)
                 return networkStream;
 
-            var expectedCert = new X509Certificate2(Convert.FromBase64String(info.Certificate), (string)null, X509KeyStorageFlags.MachineKeySet);
-            var sslStream = new SslStream(networkStream, false, (sender, actualCert, chain, errors) => expectedCert.Equals(actualCert));
+            var sslStream = new SslStream(networkStream, false, (sender, actualCert, chain, errors) =>
+            {
+                string actualStringCert = Convert.ToBase64String(actualCert.Export(X509ContentType.Cert));
```
Comment (ppekrol) [inline src/Raven.Client/Util/TcpUtils.cs]: previously we compared 2 instances of X509Certificate2, now we are comparing 2 Base64 strings. Is this what we want? Want to double check that. In all cases 'actualCert' is exportable?

File: src/Raven.Server/Utils/CertificateUtils.cs (hunk 1 above)
Comment (haludi) [reply src/Raven.Server/Utils/CertificateUtils.cs]: Yes, I missed that.
Fixed.

File: src/Raven.Client/Util/TcpUtils.cs (hunk 2 above)
Comment (haludi) [reply src/Raven.Client/Util/TcpUtils.cs]: Done

**Review by ppekrol — APPROVED** [review]

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+        var f = flag ?? FlagsForOpen;
+        try
+        {
+            return creator(f);
+        }
+        catch
+        {
+            if ((f & X509KeyStorageFlags.MachineKeySet) == X509KeyStorageFlags.MachineKeySet)
+                throw;
+            Volatile.Write(ref KeySetFlag, (int)X509KeyStorageFlags.MachineKeySet);
```
Comment (ayende) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: Why modify this value? 
Why not just always do those two tries?
Would be easier to do that and no state to manage.

Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: The error we got from the `X509Certificate2` constructor is "file not found".
This is an inner implementation error and I don't think I can rely on it.
If I don't check if I have this specific error and I just failover to the MachineKeySet we can end with sometimes loading with UserKeySet and sometimes with MachineKeySet.
I prefer it to be more consistent, so one flag will be used for all.
But I will change it to do those two tries each time if you think it is better.

Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: Done

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+    private static X509Certificate2 CreateCertificate(Func<X509KeyStorageFlags, X509Certificate2> creator, X509KeyStorageFlags? flag)
+    {
+        DebugAssertDoesntContainKeySet(flag);
+        var f = AddUserKeySet(flag);
+        try
+        {
+            return creator(f);
+        }
+        catch
+        {
```
Comment (ml054) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs:61]: can we log here? (but once)

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+    private static void Import(Action<X509KeyStorageFlags> importer, X509KeyStorageFlags? flag)
+    {
+        DebugAssertDoesntContainKeySet(flag);
+        var f = AddUserKeySet(flag);
+        try
+        {
+            importer(f);
+        }
+        catch
+        {
```
Comment (ml054) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs:30]: can we log here (but once)

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 4 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs:61]: Yes, sure.
Maybe worth to log on the Operation level for the first time and on the Information level for the rest.

Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs:61]: done

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 5 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs:30]: done

File: src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation.cs [hunk 6]
```diff
@@ -24,9 +24,9 @@ public UpdatePullReplicationAsSinkOperation(PullReplicationAsSink pullReplicatio
             if (pullReplication.CertificateWithPrivateKey != null)
             {
                 var certBytes = Convert.FromBase64String(pullReplication.CertificateWithPrivateKey);
-                using (var certificate = new X509Certificate2(certBytes,
+                using (var certificate = CertificateLoaderUtil.CreateCertificate(certBytes,
                     pullReplication.CertificatePassword,
-                    X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet))
+                    X509KeyStorageFlags.Exportable))
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation.cs]: better to use CertificateLoaderUtil.FlagsForExport

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+            if (Log.IsInfoEnabled)
+                Log.Info(CreateMsg());
+        }
+
+        string CreateMsg()
+        {
+            var msg = $"{nameof(CertificateLoaderUtil)}.{method} - Flags used {f}";
+            if (exception != null)
+                msg += $"Firs attempt exception : {exception}";
```
Comment (ppekrol) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: typo

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    {
+        const X509KeyStorageFlags keyStorageFlags = 
+            X509KeyStorageFlags.UserKeySet | 
+            X509KeyStorageFlags.MachineKeySet | 
+            (X509KeyStorageFlags)0x20 ; // X509KeyStorageFlags.EphemeralKeySet not defined in .Netstandard
+        
+        Debug.Assert(flags.HasValue == false || (flags.Value & keyStorageFlags) == 0);
+    }
+    
+    private static void LogIfNeeded(string method, X509KeyStorageFlags f, Exception exception)
```
Comment (ppekrol) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: do we really need to log each time? even of success?

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 9]
```diff
@@ -0,0 +1,119 @@
...
+        return (flag ?? X509KeyStorageFlags.DefaultKeySet) | X509KeyStorageFlags.UserKeySet;
+    }
+
+    [Conditional("DEBUG")]
+    private static void DebugAssertDoesntContainKeySet(X509KeyStorageFlags? flags)
+    {
+        const X509KeyStorageFlags keyStorageFlags = 
+            X509KeyStorageFlags.UserKeySet | 
+            X509KeyStorageFlags.MachineKeySet | 
+            (X509KeyStorageFlags)0x20 ; // X509KeyStorageFlags.EphemeralKeySet not defined in .Netstandard
```
Comment (ppekrol) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: private static X509KeyStorageFlags EphemeralKeySet =
#if NETCOREAPP_3_1_OR_GREATER
X509KeyStorageFlags.EphemeralKeySet 
#else
(X509KeyStorageFlags)0x20 // do we need this at all in that taget?
#endif

File: test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs [hunk 10]
```diff
@@ -336,7 +337,7 @@ exit 1
             X509Certificate2 serverCertificate;
             try
             {
-                serverCertificate = new X509Certificate2(certificates.ServerCertificatePath, (string)null, X509KeyStorageFlags.Exportable | X509KeyStorageFlags.MachineKeySet);
+                serverCertificate = new X509Certificate2(certificates.ServerCertificatePath, (string)null, CertificateLoaderUtil.FlagsForExport | X509KeyStorageFlags.MachineKeySet);
```
Comment (ppekrol) [inline test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs:340]: shouldn't we use the LoadUtil here?

File: test/SlowTests/Issues/RavenDB-15807.cs [hunk 11]
```diff
@@ -42,7 +43,7 @@ public async Task ChangeCertificateTypeInPullReplication()
             var hubServer = GetNewServer(new ServerCreationOptions { CustomSettings = hubSettings, RegisterForDisposal = true });
             var sinkServer = GetNewServer(new ServerCreationOptions { CustomSettings = sinkSettings, RegisterForDisposal = true });
 
-            var ownCertificate = new X509Certificate2(sinkCertificates.ClientCertificate1Path, (string)null, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable);
+            var ownCertificate = new X509Certificate2(sinkCertificates.ClientCertificate1Path, (string)null, X509KeyStorageFlags.MachineKeySet | CertificateLoaderUtil.FlagsForExport);
```
Comment (ppekrol) [inline test/SlowTests/Issues/RavenDB-15807.cs:46]: shouldn't we use the LoadUtil here?

File: test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs [hunk 12]
```diff
@@ -43,7 +45,7 @@ public async Task CanExportAndImportDatabaseRecord()
             var dummy = Certificates.GenerateAndSaveSelfSignedCertificate(createNew: true);
             string privateKey;
             using (var pullReplicationCertificate =
-                new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable))
+                new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | CertificateLoaderUtil.FlagsForExport))
```
Comment (ppekrol) [inline test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs:48]: shouldn't we use the LoadUtil here?

File: test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs [hunk 13]
```diff
@@ -249,7 +251,7 @@ public async Task CanMigrateDatabaseRecord()
...
```
Comment (ppekrol) [inline test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs:254]: shouldn't we use the LoadUtil here?

File: test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs [hunk 14]
```diff
@@ -1005,7 +1007,7 @@ public async Task CanBackupAndRestoreDatabaseRecord()
...
```
Comment (ppekrol) [inline test/SlowTests/Smuggler/BackupDatabaseRecordTests.cs:1010]: shouldn't we use the LoadUtil here?

File: test/StressTests/Issues/RavenDB-15145.cs [hunk 15]
```diff
@@ -36,7 +37,7 @@ public async Task PullReplicationWithoutPrivateKey()
             var sinkServer = GetNewServer(new ServerCreationOptions { CustomSettings = sinkSettings, RegisterForDisposal = true });
 
             var dummy = Certificates.GenerateAndSaveSelfSignedCertificate(createNew: true);
-            var pullReplicationCertificate = new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable);
+            var pullReplicationCertificate = new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | CertificateLoaderUtil.FlagsForExport);
```
Comment (ppekrol) [inline test/StressTests/Issues/RavenDB-15145.cs:40]: shouldn't we use the LoadUtil here?

File: test/StressTests/Server/Replication/PullReplicationStressTests.cs [hunk 16]
```diff
@@ -40,7 +41,7 @@ public async Task PullExternalReplicationWithCertificateShouldWork()
             var sinkServer = GetNewServer(new ServerCreationOptions { CustomSettings = sinkSettings, RegisterForDisposal = true });
 
             var dummy = Certificates.GenerateAndSaveSelfSignedCertificate(createNew: false);
-            var pullReplicationCertificate = new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.Exportable);
+            var pullReplicationCertificate = new X509Certificate2(dummy.ServerCertificatePath, (string)null, X509KeyStorageFlags.MachineKeySet | CertificateLoaderUtil.FlagsForExport);
```
Comment (ppekrol) [inline test/StressTests/Server/Replication/PullReplicationStressTests.cs:44]: shouldn't we use the LoadUtil here?

**Comment by ppekrol** [comment]
Are there any backward compatibility concerns here? Previously certs were written in MachineKeySet, now in UserKeySet. If I have an old machine, is there sth that we should take into account?

File: test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs (hunk 10 above)
Comment (haludi) [reply test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs:340]: As I see it, this is a test so there is no need to change it.

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 9 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: done

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 8 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: If there will be an issue in this area I would like to see if there were cases where the certificate was loaded in UserKeySet.
This is only an information level

File: src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation.cs (hunk 6 above)
Comment (haludi) [reply src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation.cs]: done

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 7 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: done

**Comment by haludi** [comment]
> src/Raven.Client/Documents/Operations/Replication/UpdatePullReplicationAsSinkOperation.cs

Before we changed the flag to MachineKeySet we used the default which is usually UserKeySet
> DefaultKeySet | 0 | The default key set is used. The user key set is usually the default.
https://learn.microsoft.com/en-us/dotnet/api/system.security.cryptography.x509certificates.x509keystorageflags?view=net-6.0#fields

The idea is to failover if there is an issue.

File: test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs (hunk 10 above)
Comment (ppekrol) [reply test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs:340]: but it generates extra files in windows machine key sets, because of this, right?

Comment (haludi) [reply test/SlowTests/ExtensionPoints/ExtensionPointsTests.cs:340]: What do you mean extra?
It creates the temp private key file in the machine folder and creates the public key in the Registry

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 17]
```diff
@@ -85,10 +85,12 @@ private static X509KeyStorageFlags AddUserKeySet(X509KeyStorageFlags? flag)
     [Conditional("DEBUG")]
     private static void DebugAssertDoesntContainKeySet(X509KeyStorageFlags? flags)
     {
-        const X509KeyStorageFlags keyStorageFlags = 
-            X509KeyStorageFlags.UserKeySet | 
-            X509KeyStorageFlags.MachineKeySet | 
-            (X509KeyStorageFlags)0x20 ; // X509KeyStorageFlags.EphemeralKeySet not defined in .Netstandard
+        const X509KeyStorageFlags keyStorageFlags =
+#if NETCOREAPP_3_1_OR_GREATER
```
Comment (ppekrol) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: it looks like the key is NETCOREAPP3_1_OR_GREATER (VS should give you intellisense on that)

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 8 above)
Comment (ppekrol) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: Isn't that an overkill? If you failed once for UserKeySet, then I'm expecting that you will fail each time there anyway?

File: src/Raven.Client/Util/CertificateLoaderUtil.cs (hunk 17 above)
Comment (haludi) [reply src/Raven.Client/Util/CertificateLoaderUtil.cs]: done

**Review by grisha-kotler — APPROVED** [review]

File: src/Raven.Client/Util/CertificateLoaderUtil.cs [hunk 18]
```diff
@@ -0,0 +1,133 @@
...
+                Log.Info(CreateMsg());
+        }
+
+        string CreateMsg()
+        {
+            var msg = $"{nameof(CertificateLoaderUtil)}.{method} - Flags used {f}";
+            if (exception != null)
+                msg += $"First attempt exception : {exception}";
+            return msg;
+        }
```
Comment (ayende) [inline src/Raven.Client/Util/CertificateLoaderUtil.cs]: We have support for outputting the exception message _separately_ for a reason.
For example, it indicates an `error` when using debug log viewer in the studio, let's use the API properly

**Review by ayende — CHANGES_REQUESTED** [review]

**Review by ayende — APPROVED** [review]

**Pull request merged**